DevTools/SourceMaps/explainer.md

```diff
@@ -0,0 +1,245 @@
+
+We want to adopt the general approach taken by the `pasta-sourcemaps` ("Pretty (and) Accurate Stack Trace Analysis") library. However, instead of adding to the `names` field, we will add two fields.
+
+* Add two additional field to the source map: `scopes` and `scopeNames`
+* The `scopes` field should be a list of ranges and pointers into `scopeNames`
+* The ranges in the field should always point to **Original Source** locations.
+
+### Practical example
+
+Suppose we have the following TypeScript source:
+
+```ts
+// example.ts
+1 function foo(bar: number): never {
+2     throw new Error('Intentional.');
+3 }
+4 foo();
+```
+
+Direct TypeScript compilation would emit something fairly similar:
+
+```js
+// example.js
+1 function foo(bar) {
+2     throw new Error('Intentional.');
+3 }
+4 foo();
+```
+
+But running it through a minifier would probably transform it to an IIFE lambda:
+
+```js
+// example.min.js
+1 (()=>{throw new Error('Intentional.')})();
+```
+
+Examining the call stack of this looks something like:
+
+```
+Error: Intentional.
+    at <anonymous>:1:13
+    at <anonymous>:1:40
+```
+
+We can get to the exact *locations* with source maps today, but assuming that the `mapping`'s 5th field ("the zero-based index into the `names` list associated with the segment) is encoded as it typically is, which is meant for live debugging, the decoded stack would actually look like this:
+
+```
+Error: Intentional
+    at throw (example.js:2:5)
+    at foo (example.js:4:1)
+```
+
+What we want to get to from here would be, at minimum:
+
+```
+Error: Intentional.
+    at foo (example.js:2:5)
+    at Global code (example.js:4:1)
+```
+
+### Improving the contents of the map
+
+Let us re-examine the original code in this example:
+
+```ts
+1 function foo(bar: number): never {
+2     throw new Error('Intentional.');
+3 }
+4 foo();
+```
+
+The scopes list should be as follows:
+
+1. Line 1 Col 1 -> Line 1 Col 34 (captures `function foo(bar: number): never ` including the trailing space): `Global code`
+1. Line 1 Col 35 -> Line 3 Col 1 (captures the entire body of `foo`): `foo`
+1. Line 3 Col 2 -> Line 4 Col 7 (captures the newline + invocation of `foo`): `Global code`
+
+By de-duplicating the names, that yields the expected `scopeNames` value of `['Global code', 'foo']`.
+
+Finally, `scopes` will need to be an array of strings, and this array should have the same number of entries as the `sources` field. Each string will be a comma-delimited list of Base64-encoded VLQs with the following set of entries (please note, all offsets are 0-based, in concurrence with the previous Source Maps spec):
+
+1. A value representing starting line, relative to the previous starting line value. If this is the first value, it is an absolute value.
+2. A value representing the absolute starting column number.
+3. A value representing the ending line, **relative to the starting line** value.
+4. A value representing the ending starting column number.

```

Comment:
There may be similar considerations for the other fields too, I just opted to go a bit deeper on this one in particular

Reply:
Relative ending column may be preferable here to absolute. Now that I have the TypeScript compiler change, I'm trying to get it to compile itself to see how the performance characteristics change, and we can use iteration here to check which produces smaller values. My supposition, however, is that most often, the absolute values of ending column will generally be < the absolute values of the starting column, because that's usually how we write JS:

```js
function doSomething() { // <-- starting column
 // ...
} // <-- ending column
```

That's less true for transforming already-minified files, e.g., react-dom.production.min.js.

I'll do some experiments with the TypeScript compiler and report back before we call this "completed"